src/mds/Server.cc

```diff
@@ -2139,7 +2199,40 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
 
   if (mdr->killed) {
     dout(10) << "request " << *mdr << " was killed" << dendl;
-    return;
+    //if the mdr is a "batch_op" and it has followers, pick a follower as
+    //the new "batch_head" and go on processing the new one.
+    if (mdr->is_batch_op() && mdr->is_batch_head && mdr->batch_reqs.size()) {
+      auto new_batch_head = *(mdr->batch_reqs.begin());
+      mdr->batch_reqs.erase(new_batch_head);
+      for (auto r : mdr->batch_reqs) {
+	r->batch_head = new_batch_head;
+	new_batch_head->batch_reqs.insert(r);
+	mdr->batch_reqs.erase(r);
+      }
+      new_batch_head->batch_head.reset();
+      mdr = new_batch_head;
+      mdr->is_batch_head = true;
+      int mask = mdr->client_request->head.args.getattr.mask;
+
+      if (mdr->client_request->get_op() == CEPH_MDS_OP_GETATTR) {
+	auto fin = mdr->in[0]->batch_ops[mask];
+	fin->set_request(new_batch_head);
+      } else if (mdr->client_request->get_op() == CEPH_MDS_OP_LOOKUP) {
+	auto fin = mdr->dn[0].back()->batch_ops[mask];
+	fin->set_request(new_batch_head);
+      }
+      mdr->killed = false;
```

Comment:
you can't blindly mdr->reset killed.  Othwerwise, looks good

Reply:
@ukernel Hi, thanks for the review:-)
Could you provide some more details about why can't directly reset mdr->killed here? I'm asking this because the mdr has been completely reset to new_batch_head, which is a completely different mdr now. Thanks:-)